### PR TITLE
Prevent misoptimisation in `StringOps`.

### PR DESCRIPTION
**Pull request by max-schaefer**
This is a regression relative to 1.24, hence a hotfix.

### CODE REVIEW

File: ql/src/semmle/go/StringOps.qll [hunk 1]
```diff
@@ -78,15 +78,25 @@ module StringOps {
...
+    pragma[noinline]
+    private predicate comparesToZero(DataFlow::EqualityTestNode eq, DataFlow::Node nd) {
+      exists(DataFlow::Node zero |
+        eq.hasOperands(globalValueNumber(nd).getANode(), zero) and
+        zero.getIntValue() = 0
+      )
+    }
+
+    /**
+     * An expression of form `strings.Index(A, B) == 0`.
```
Comment (intrigus-lgtm) [inline ql/src/semmle/go/StringOps.qll:92]:
```suggestion
     * An expression of the form `strings.Index(A, B) == 0`.
```

File: ql/src/semmle/go/StringOps.qll [hunk 2]
```diff
@@ -97,19 +107,30 @@ module StringOps {
...
+      exists(DataFlow::ElementReadNode read |
+        eq.hasOperands(globalValueNumber(read).getANode(), rhs) and
+        str = read.getBase() and
+        str.getType().getUnderlyingType() instanceof StringType and
+        read.getIndex().getIntValue() = 0
+      )
+    }
+
+    /**
+     * A comparison of form `x[0] == 'k'` for some rune literal `k`.
```
Comment (intrigus-lgtm) [inline ql/src/semmle/go/StringOps.qll:125]:
```suggestion
     * A comparison of the form `x[0] == 'k'` for some rune literal `k`.
```

File: ql/src/semmle/go/StringOps.qll (hunk 1 above)
Comment (max-schaefer) [reply ql/src/semmle/go/StringOps.qll:92]: Thanks!

File: ql/src/semmle/go/StringOps.qll [hunk 3]
```diff
@@ -97,27 +107,38 @@ module StringOps {
...
-      DataFlow::ElementReadNode read;
+      DataFlow::Node base;
       DataFlow::Node runeLiteral;
 
-      HasPrefix_FirstCharacter() {
-        read.getBase().getType().getUnderlyingType() instanceof StringType and
-        read.getIndex().getIntValue() = 0 and
-        eq(_, globalValueNumber(read).getANode(), runeLiteral)
-      }
+      HasPrefix_FirstCharacter() { comparesFirstCharacter(this, base, runeLiteral) }
```
Comment (sauyon) [inline ql/src/semmle/go/StringOps.qll:131]: Just out of interest, how is this different to inlining the predicate here?

Comment (max-schaefer) [reply ql/src/semmle/go/StringOps.qll:131]: Apparently it affects how magic sets is applied :man_shrugging:.

**Pull request merged**